src/curl.c

```diff
@@ -539,7 +526,107 @@ static int swupd_curl_get_file_full(const char *url, char *filename,
 		unlink(filename);
 	}
 
-	return err;
+	return -status;
+}
+
+/*
+ * Determines what strategy to use based on the return code:
+ * - do not retry: if the fault indicates that the failure isn't transient or
+ *                 is unlikely to be successful if repeated (i.e. disk full)
+ * - retry immediately: if the specific fault reported is unusual or rare, it
+ *                      might have been caused by unusual circumstances, the
+ *                      same failure is unlikely to be repeated (e.g. corrupt
+ *                      network package)
+ * - retry after a delay: if the fault is caused by a transient fault (like
+ *                        network connectivity)
+ */
+static enum retry_strategy determine_strategy(int ret_code)
+{
+	switch (ret_code) {
+	case DOWNLOAD_STATUS_FORBIDDEN:
+	case DOWNLOAD_STATUS_NOT_FOUND:
+	case DOWNLOAD_STATUS_WRITE_ERROR:
+		return DONT_RETRY;
+	case DOWNLOAD_STATUS_RANGE_ERROR:
+	case DOWNLOAD_STATUS_PARTIAL_FILE:
+		return RETRY_NOW;
+	case DOWNLOAD_STATUS_ERROR:
+	case DOWNLOAD_STATUS_TIMEOUT:
+		return RETRY_WITH_DELAY;
+	default:
+		return RETRY_NOW;
+	}
+}
+
+static int retry_download(const char *url, char *filename, struct curl_file_data *in_memory_file, bool resume_ok)
+{
+
+	int current_retry = 0;
+	int sleep_time = retry_delay;
+	int strategy;
+	int ret;
+
+	for (;;) {
+
+		/* download file */
+		ret = swupd_curl_get_file_full(url, filename, in_memory_file, resume_ok);
+
+		if (ret == DOWNLOAD_STATUS_COMPLETED) {
+			/* operation successful */
+			break;
+		} else {
```

Comment:
It's easier to read if you take this else out because break is going to get out of the for.

Reply:
I'm not sure I understand what you mean on this one

Reply:
can you elaborate a little more please?

Reply:
No. You just don't need this else here and it's easier to read because the code won't have an extra indentation

    for (;;) {
             ret = swupd_curl_get_file_full(url, filename, in_memory_file, resume_ok);
            if (ret == DOWNLOAD_STATUS_COMPLETED) {
                break; //Or you can add a return here instead of the break
            }
             /*code that is in the else come here*
    }

Reply:
done